### ISSUE TITLE
Structs that use __attribute__((aligned(sizeof(...)))) cause invalid JSON to be generated

### ISSUE THREAD

**Issue by alexer** [iovisor/bcc#2605]
The following code leads to `json.decoder.JSONDecodeError` in `BPF.get_table()`:
```
from bcc import BPF

b = BPF(text="""
struct S {
        int x;
} __attribute__((aligned(sizeof(long))));

BPF_HISTOGRAM(typetest, struct S);
""")

typetest = b.get_table("typetest")
print(typetest.Key._fields_)
```

The resulting JSON looks like this:
`["S", [["x","int"], ["__pad_end","char",[4]] ], "struct_packed"]"long"`

It seems that any struct that uses `sizeof()` in `__attribute__((aligned(...)))` causes the size to get inserted into the JSON. (For searchability: If the attribute is on the outermost struct, the error message will be `Extra data`, otherwise it'll be `Expecting ',' delimiter`)

**Comment by yonghong-song** [comment]
What is exactly the failed test case?

The above code seems working fine for me.
```
bash-4.4$ cat t.py
#!/usr/bin/python

from bcc import BPF

b = BPF(text="""
struct S {
        int x;
} __attribute__((aligned(sizeof(long))));

BPF_HISTOGRAM(typetest, struct S);
""")

typetest = b.get_table("typetest")
print(typetest.Key._fields_)

-bash-4.4$ sudo ./t.py
[(u'x', <class 'ctypes.c_int'>), (u'__pad_end', <class 'bcc.c_char_Array_4'>)]
-bash-4.4$
```

The `_pad_*` is added to make padding explicit. The original reason is to make padding explicit, esp. regarding to __int128 type so that C type and python ctype will see the same layout.

The `aligned(...)` attribute will affect alignment, hence possible structure size and padding.
The statement `causes the size to get inserted into the JSON` is not accurate.

**Comment by alexer** [comment]
Hm, that's weird; that exact code does not work for me:
```
odoryuu ~ # cat bcc_structinfo3.py 
from bcc import BPF

b = BPF(text="""
struct S {
        int x;
} __attribute__((aligned(sizeof(long))));

BPF_HISTOGRAM(typetest, struct S);
""")

typetest = b.get_table("typetest")
print(typetest.Key._fields_)

odoryuu ~ # python bcc_structinfo3.py 
Traceback (most recent call last):
  File "bcc_structinfo3.py", line 11, in <module>
    typetest = b.get_table("typetest")
  File "/usr/lib64/python3.6/site-packages/bcc/__init__.py", line 485, in get_table
    keytype = BPF._decode_table_type(json.loads(key_desc))
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 342, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 1 column 65 (char 64)
```

I ran the test on Gentoo, with these package versions: bcc-0.10.0 clang-7.1.0 llvm-7.1.0
(I did also try the master branch of bcc from github, and got the same error - though I didn't `make install`, I just ran the script so it had access to the new version, and verified that it used the correct version with `strace`; I also tried it with python-2.7, but got the same result)

Is there any other relevant information or package versions that I could provide?

**Comment by yonghong-song** [comment]
I am using python2 in my local environment which cannot reproduce this issue. Maybe this is the issue? I need to rebuild my python3 based bcc to give a try in the next couple of days.

**Comment by alexer** [comment]
I did reproduce it on python2 too, though. Maybe it could be some difference between llvm/clang versions?

FYI, I also tried it on Ubuntu 18.04; I could reproduce it there as well with bcc from git. (The latest bpfcc-tools version is 0.5.0-5ubuntu1; it worked on that, but I think that's just because the version is from before the explicit padding code was introduced) The version for llvm and clang was 1:6.0-41\~exp5\~ubuntu1 on that machine.

I had previously tracked the extraneous output to `BMapDeclVisitor::VisitBuiltinType`, but hadn't been able to figure out how to get clang to dump the AST for the RecordDecl (I'm not really that familiar with either C++ or clang), so the only info I'd managed to squeeze out was stacktraces of calls to `VisitBuiltinType`; I just figured out how to do the dumping though, so:

When I add `D->dump();` to `BMapDeclVisitor::VisitRecordDecl`, I get:
```
RecordDecl 0x5642d3085ae8 </virtual/main.c:2:1, line:4:1> line:2:8 struct S definition
|-AlignedAttr 0x5642d3085c28 <line:4:18, col:38> aligned
| `-UnaryExprOrTypeTraitExpr 0x5642d3085c08 <col:26, col:37> 'unsigned long' sizeof 'long'
`-FieldDecl 0x5642d3085ba8 <line:3:2, col:6> col:6 x 'int'
```
I'm not sure if the AST dump is 1:1 with the actual nodes, since the stacktraces for calls to `BMapDeclVisitor::VisitBuiltinType` show deeper nesting:

This seems to be the stacktrace when it's called for the type of `x`:
```
#0  0x00007ffff28add50 in ebpf::BMapDeclVisitor::VisitBuiltinType(clang::BuiltinType const*)@plt () from ./libbcc.so.0
#1  0x00007ffff29777e6 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseTypeLoc(clang::TypeLoc) () from ./libbcc.so.0
#2  0x00007ffff2982bab in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseFieldDecl(clang::FieldDecl*) () from ./libbcc.so.0
#3  0x00007ffff296cbb8 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseDecl(clang::Decl*) () from ./libbcc.so.0
#4  0x00007ffff296afe2 in ebpf::BMapDeclVisitor::genJSONForField(clang::FieldDecl*) () from ./libbcc.so.0
#5  0x00007ffff296b954 in ebpf::BMapDeclVisitor::VisitRecordDecl(clang::RecordDecl*) () from ./libbcc.so.0
#6  0x00007ffff296d377 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseDecl(clang::Decl*) () from ./libbcc.so.0
#7  0x00007ffff2976438 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseType(clang::QualType) () from ./libbcc.so.0
#8  0x00007ffff297655a in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseType(clang::QualType) () from ./libbcc.so.0
#9  0x00007ffff29769ee in ebpf::JsonMapTypesVisitor::Visit(ebpf::TableDesc&, clang::ASTContext&, clang::QualType, clang::QualType) () from ./libbcc.so.0
```

And this seems to be the stacktrace when it's called for the type in the `sizeof()`:
```
#0  0x00007ffff28add50 in ebpf::BMapDeclVisitor::VisitBuiltinType(clang::BuiltinType const*)@plt () from ./libbcc.so.0
#1  0x00007ffff29777e6 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseTypeLoc(clang::TypeLoc) () from ./libbcc.so.0
#2  0x00007ffff297a33d in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseUnaryExprOrTypeTraitExpr(clang::UnaryExprOrTypeTraitExpr*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) () from ./libbcc.so.0
#3  0x00007ffff296c06d in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseStmt(clang::Stmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) [clone .constprop.2734] () from ./libbcc.so.0
#4  0x00007ffff296c444 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseDecl(clang::Decl*) () from ./libbcc.so.0
#5  0x00007ffff2976438 in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseType(clang::QualType) () from ./libbcc.so.0
#6  0x00007ffff297655a in clang::RecursiveASTVisitor<ebpf::BMapDeclVisitor>::TraverseType(clang::QualType) () from ./libbcc.so.0
#7  0x00007ffff29769ee in ebpf::JsonMapTypesVisitor::Visit(ebpf::TableDesc&, clang::ASTContext&, clang::QualType, clang::QualType) () from ./libbcc.so.0
```

**Comment by yonghong-song** [comment]
The above recursiveASTVisit is normal as AST traversal going from type to decl/stmt to record declaration and fields.

Could you print `key_desc` as in  the error stack 
`keytype = BPF._decode_table_type(json.loads(key_desc))`?
This may help to find out what is the likely problem.

**Comment by yonghong-song** [comment]
okay. Tried llvm 8. It failed with the error you described the above. It looks during parsing,
the AST visitor tries to parse the type 'long' in sizeof() attribute (as you described in the above)
so the final key_desc is
```
["S", [["x","int"], ["__pad_end","char",[4]] ], "struct_packed"]"long"
```

While with llvm 10, the key_desc is
```
["S", [["x","int"], ["__pad_end","char",[4]] ], "struct_packed"].
```
To workaround the issue, just use constant instead of sizeof() like
```
struct S {
        int x;
} __attribute__((aligned(8)));
```

Do you think this workaround is good enough for you? Considering newer compiler is okay for this, I intend to leave it as is. But do let me know if you have a strong feeling that we should fix this for llvm 8.

**Comment by alexer** [comment]
I had it in the first message, it's `["S", [["x","int"], ["__pad_end","char",[4]] ], "struct_packed"]"long"`.

The problem seems to be that after it's built the correct JSON, it appends the type in the `sizeof()` to the end. The extraneous `"long"` is added from `BMapDeclVisitor::VisitBuiltinType`, which is called twice - once for the type of `x`, and in the end for the type in the `sizeof()`.

I assume the second call comes when the `RecursiveASTVisitor` handles the `AlignedAttr`/`UnaryExprOrTypeTraitExpr` seen in the AST. Since it works for you, I think there's some difference between the AST that clang generates for you and me.

In other words, in addition to being called as intended during `genJSONForField`, `VisitBuiltinType` is also being called outside of that, from the part of the AST dealing with the `__alignment__`. (Maybe the right thing to do could be to define a `VisitAlignedAttr` that doesn't traverse its children?)

**Comment by yonghong-song** [comment]
> Maybe the right thing to do could be to define a VisitAlignedAttr that doesn't traverse its children?
This makes sense, we should not traverse types for any of its attributes.
Could you try and submit a pull request for this with a test case? I will test with newer llvm. Our buildbot should test old llvm's.
Thanks!

**Comment by yonghong-song** [comment]
> I had it in the first message, it's ["S", [["x","int"], ["__pad_end","char",[4]] ], "struct_packed"]"long".
Sorry about this. I emphasized to reproduce the issue and did not pay attention to the analysis part ...

**Comment by alexer** [comment]
I just tested what I suggested, and it seems to work - but since you said "for any of its attributes", I think maybe it should be `VisitAttr` instead - that seems to work too.

> Could you try and submit a pull request for this with a test case?

Yeah, sure!

> Sorry about this. I emphasized to reproduce the issue and did not pay attention to the analysis part ...

No problem at all :)
